aws_lambda_powertools/event_handler/api_gateway.py

```diff
@@ -153,58 +172,62 @@ def register_resolver(func: Callable):
-        if route is None:  # No matching route was found
-            return response.to_dict()
-
-        if route.cors:
-            response.add_cors(self._cors or CORSConfig())
-        if route.cache_control:
-            response.add_cache_control(route.cache_control)
-        if route.compress and "gzip" in (self.current_event.get_header_value("accept-encoding") or ""):
-            response.compress()
+        return self._resolve_response().build(self.current_event, self._cors)
 
-        return response.to_dict()
+    def __call__(self, event, context) -> Any:
+        return self.resolve(event, context)
 
     @staticmethod
     def _compile_regex(rule: str):
+        """Precompile regex pattern"""
         rule_regex: str = re.sub(r"(<\w+>)", r"(?P\1.+)", rule)
         return re.compile("^{}$".format(rule_regex))
 
     def _to_data_class(self, event: Dict) -> BaseProxyEvent:
+        """Convert the event dict to the corresponding data class"""
         if self._proxy_type == ProxyEventType.http_api_v1:
             return APIGatewayProxyEvent(event)
         if self._proxy_type == ProxyEventType.http_api_v2:
             return APIGatewayProxyEventV2(event)
         return ALBEvent(event)
 
-    def _find_route(self, method: str, path: str) -> Tuple[Optional[Route], Response]:
+    def _resolve_response(self) -> ResponseBuilder:
+        """Resolve the response or return the not found response"""
+        method = self.current_event.http_method.upper()
+        path = self.current_event.path
         for route in self._routes:
             if method != route.method:
                 continue
             match: Optional[re.Match] = route.rule.match(path)
             if match:
                 return self._call_route(route, match.groupdict())
 
+        return self._not_found(method, path)
+
+    def _not_found(self, method: str, path: str) -> ResponseBuilder:
+        """No matching route was found, includes support for the cors preflight response"""
         headers = {}
         if self._cors:
             headers.update(self._cors.to_dict())
             if method == "OPTIONS":  # Preflight
                 headers["Access-Control-Allow-Methods"] = ",".join(sorted(self._cors_methods))
-                return None, Response(status_code=204, content_type=None, body=None, headers=headers)
-
-        return None, Response(
-            status_code=404,
-            content_type="application/json",
-            body=json.dumps({"message": f"No route found for '{method}.{path}'"}),
-            headers=headers,
+                return ResponseBuilder(Response(status_code=204, content_type=None, body=None, headers=headers))
+        return ResponseBuilder(
+            Response(
+                status_code=404,
+                content_type="application/json",
+                body=json.dumps({"message": f"No route found for '{method}.{path}'"}),
```

Comment:
**Note to self**: Pull the entire code to check whether we're not making customers vulnerable to data enumeration attacks via HTTP status code